### PR TITLE
tropical planet

### PR DESCRIPTION
**Pull request by jschroettle**
Hi Yair,
just added the changes we did together to Forcing.pyx, Forcing.pxd and the c-kernel
When you review you it, I will run it after the weekend

### CODE REVIEW

File: Forcing.pyx [hunk 1]
```diff
@@ -140,6 +142,59 @@ cdef class HelzSuarez(ForcingBase):
...
+			Py_ssize_t i,j
+			Py_ssize_t nx = Pr.nlats
+			Py_ssize_t ny = Pr.nlons
+		self.Tbar = np.zeros((Pr.nlats, Pr.nlons, Pr.n_layers), dtype=np.float64, order='c')
+		self.sin_lat = np.zeros((Pr.nlats, Pr.nlons), dtype=np.float64, order='c')
+		self.cos_lat = np.zeros((Pr.nlats, Pr.nlons), dtype=np.float64, order='c')
+		with nogil:
+			for i in range(nx):
+				for j in range(ny):
+					self.sin_lat[i,j] = sin(Gr.lat[i,j])
```
Comment (yairchn) [inline Forcing.pyx]: this is not needed in the TP forcing

File: Forcing.pyx [hunk 2]
```diff
@@ -140,6 +142,59 @@ cdef class HelzSuarez(ForcingBase):
...
+					self.cos_lat[i,j] = cos(Gr.lat[i,j])
```
Comment (yairchn) [inline Forcing.pyx]: this is not needed

File: Forcing.pyx [hunk 3]
```diff
@@ -140,6 +142,59 @@ cdef class HelzSuarez(ForcingBase):
...
+		ForcingBase.__init__(self, namelist)
+		return
+
+	cpdef initialize(self, Parameters Pr, Grid Gr, namelist):
+		cdef:
+			Py_ssize_t i,j
+			Py_ssize_t nx = Pr.nlats
+			Py_ssize_t ny = Pr.nlons
+		self.Tbar = np.zeros((Pr.nlats, Pr.nlons, Pr.n_layers), dtype=np.float64, order='c')
+		self.sin_lat = np.zeros((Pr.nlats, Pr.nlons), dtype=np.float64, order='c')
```
Comment (yairchn) [inline Forcing.pyx]: this is not needed

File: Forcing.pyx [hunk 4]
```diff
@@ -140,6 +142,59 @@ cdef class HelzSuarez(ForcingBase):
...
+		self.cos_lat = np.zeros((Pr.nlats, Pr.nlons), dtype=np.float64, order='c')
```
Comment (yairchn) [inline Forcing.pyx]: this is not needed

File: Forcing.pyx [hunk 5]
```diff
@@ -140,6 +142,59 @@ cdef class HelzSuarez(ForcingBase):
...
+		cdef:
+			Py_ssize_t nx = Pr.nlats
+			Py_ssize_t ny = Pr.nlons
+			Py_ssize_t nl = Pr.n_layers
+
+		with nogil:
+			focring_tp(Pr.kappa, Pr.p_ref, Pr.sigma_b, Pr.k_a, Pr.k_b, Pr.k_f, Pr.k_s,
+						Pr.Dtheta_z, Pr.T_equator, Pr.DT_y, &PV.P.values[0,0,0],
+						&PV.T.values[0,0,0],&self.Tbar[0,0,0], &self.sin_lat[0,0],
+						&self.cos_lat[0,0], &DV.U.values[0,0,0], &DV.V.values[0,0,0],
```
Comment (yairchn) [inline Forcing.pyx]: you can remove "&self.sin_lat[0,0]," and "&self.cos_lat[0,0]" from the input

File: Forcing.pyx [hunk 6]
```diff
@@ -140,6 +142,59 @@ cdef class HelzSuarez(ForcingBase):
...
+	@cython.boundscheck(False)
+	@cython.cdivision(True)
+	cpdef update(self, Parameters Pr, Grid Gr, PrognosticVariables PV, DiagnosticVariables DV):
+		cdef:
+			Py_ssize_t nx = Pr.nlats
+			Py_ssize_t ny = Pr.nlons
+			Py_ssize_t nl = Pr.n_layers
+
+		with nogil:
+			focring_tp(Pr.kappa, Pr.p_ref, Pr.sigma_b, Pr.k_a, Pr.k_b, Pr.k_f, Pr.k_s,
```
Comment (yairchn) [inline Forcing.pyx]: you need to add 'forcing_tp' as external function like 'forcing_hs' is add in line 21 :
'void focring_hs(double kappa, double p_ref, double'.
Make sure the input to forcing_tp matches in the definitions

File: c_kernels/forcing_functions.h [hunk 7]
```diff
@@ -70,4 +70,72 @@ void focring_hs(
...
+           double k_b,
+           double k_f,
+           double k_s,
+           double Dtheta_z,
+           double T_equator,
+           double DT_y,
+           double* restrict p,
+           double* restrict T,
+           double* restrict T_bar,
+           double* restrict sin_lat,
```
Comment (yairchn) [inline c_kernels/forcing_functions.h]: remove this from input

File: c_kernels/forcing_functions.h [hunk 8]
```diff
@@ -70,4 +70,72 @@ void focring_hs(
...
+           double* restrict cos_lat,
```
Comment (yairchn) [inline c_kernels/forcing_functions.h]: remove this from input

**Review by yairchn — COMMENTED** [review]
You might need to pull master into your branch first, make sure this branch is committed and do 
"git checkout master"
"git pull yair master" 
git checkout "yourPR branch" 
git merge master
fix conflicts and commit again

**Comment by jschroettle** [comment]
included your comments yair + removed the truncation in diffusion like you did. will run a first trial after the weekend and send you first plots of the new temperature fields

File: Forcing.pyx [hunk 9]
```diff
@@ -18,18 +18,27 @@ from libc.math cimport pow, log, sin, cos, fmax
...
+
+
 
 def ForcingFactory(namelist):
 	if namelist['forcing']['forcing_model'] == 'None':
 		return ForcingNone(namelist)
 	elif namelist['forcing']['forcing_model'] == 'HeldSuarez':
 		return HelzSuarez(namelist)
+	elif namelist['forcing']['forcing_model'] == 'TropicalPlanet':
+		return HelzSuarez(namelist)
```
Comment (yairchn) [inline Forcing.pyx]: change to 'return TropicalPlanet'

File: HeldSuarez.in [hunk 10]
```diff
@@ -44,7 +44,13 @@
         "number_of_latitute_points": 64,
         "number_of_longitude_points": 128,
         "pressure_levels": [
+<<<<<<< HEAD
```
Comment (yairchn) [inline HeldSuarez.in]: you did not solve the conflict

File: HeldSuarez.in [hunk 11]
```diff
@@ -44,7 +44,13 @@
...
+            20000.0,
+            30000.0,
+            40000.0,
+=======
```
Comment (yairchn) [inline HeldSuarez.in]: you did not solve the conflict

File: HeldSuarez.in [hunk 12]
```diff
@@ -44,7 +44,13 @@
...
             25000.0,
+>>>>>>> df92f004ac84edf77e8844fbf29850d5d4e030ce
```
Comment (yairchn) [inline HeldSuarez.in]: you did not solve the conflict

File: HeldSuarez.in [hunk 13]
```diff
@@ -56,9 +62,15 @@
         "noise_amplitude": 1,
         "noise_type": "red",
         "restart": false,
+<<<<<<< HEAD
```
Comment (yairchn) [inline HeldSuarez.in]: you did not solve the conflict

File: HeldSuarez.in [hunk 14]
```diff
@@ -56,9 +62,15 @@
...
+        "restart folder": "/Output.HeldSuarez.48-test9lay/",
+        "restart time": "864000",
+        "restart type": "3D_output"
+=======
```
Comment (yairchn) [inline HeldSuarez.in]: you did not solve the conflict

File: HeldSuarez.in [hunk 15]
```diff
@@ -68,7 +80,11 @@
     "meta": {
         "casename": "HeldSuarez",
         "simname": "HeldSuarez",
+<<<<<<< HEAD
```
Comment (yairchn) [inline HeldSuarez.in]: you did not solve the conflict

File: HeldSuarez.in [hunk 16]
```diff
@@ -68,7 +80,11 @@
...
+        "uuid": "b8787bf4-7bbe-4608-8819-.48-test9lay"
+=======
```
Comment (yairchn) [inline HeldSuarez.in]: you did not solve the conflict

File: HeldSuarez.in [hunk 17]
```diff
@@ -68,7 +80,11 @@
...
         "uuid": "f6bfc4a1-e1d4-48a3-b6e8-no_truncation"
+>>>>>>> df92f004ac84edf77e8844fbf29850d5d4e030ce
```
Comment (yairchn) [inline HeldSuarez.in]: you did not solve the conflict

File: scripts/order.sh [hunk 18]
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash
```
Comment (yairchn) [inline scripts/order.sh:1]: you need to remove this script

File: viz/plot_ek.py [hunk 19]
```diff
@@ -109,6 +111,47 @@ def keSpectral_flux(u,v):
 
     return [Ek_sum,k]
```
Comment (yairchn) [inline viz/plot_ek.py:113]: great that you added this to the repo!

File: viz/profiles_Wp_one.py [hunk 20]
```diff
@@ -26,18 +27,18 @@ def main():
...
-    fig,ax = plt.subplots(1,3, sharey='all',figsize=(16, 7),squeeze=False)
+    fig,ax = plt.subplots(1,8, sharey='all',figsize=(16, 7),squeeze=False)
 
     i = 0
-    li=[4,3,2,1,.5]
-    al=[0.4,0.6,0.8,0.9,1.]
-    #la=['125 hPa','250 hPa','500 hPa','750 hPa','850 hPa']
+    li=[2,2,2,2,2,2,2,2]
+    al=[0.3,0.4,0.5,0.6,0.7,0.8,0.9,1.]
+    la=['200 hPa','300 hPa','400 hPa','500 hPa','600 hPa','700 hPa','800 hPa','900 hPa']
```
Comment (yairchn) [inline viz/profiles_Wp_one.py:38]: we could read the pressures from the stats file instead

**Review by yairchn — COMMENTED** [review]
there are some things left from conflict when you pulled my master, 
other comments are also there but I think its the last round

File: viz/profiles_Wp_one.py (hunk 20 above)
Comment (jschroettle) [reply viz/profiles_Wp_one.py:38]: yes, we should - do you have an idea how to retrieve those when we know the namelist? would make plotting a lot easier

File: scripts/order.sh (hunk 18 above)
Comment (jschroettle) [reply scripts/order.sh:1]: i currently use this to merge temperature files for plotting wheeler kiladis

File: Forcing.pyx (hunk 9 above)
Comment (jschroettle) [reply Forcing.pyx]: good catch, i included the StochasticTropicalPlanet now and also StochasticHeldSuarez

in the tp case, we won't run one without Stochastic forcing, i believe

**Comment by jschroettle** [comment]
Hi yair, remaining conflicts are solved- I need to run a test case now

**Comment by yairchn** [comment]
> Hi yair, remaining conflicts are solved- I need to run a test case now

please do run it from this branch.
If it runs without breaking - irrespective of the results - I will merge the PR

**Comment by jschroettle** [comment]
.... running ... yair!

**Pull request merged**